# **Topic modelling**
Topic modelling using the Latent Dirichlet Allocation (LDA)

## description
Topic modelling aims to discover the hidden semantic structures of a large text
corpus, with numerous applications such as automatic categorisation of documents,
text mining, text information retrieval, to name a few.

## implementation

following steps to are being followed in implementation:--

1. get the sample data
2. clean the data
3. use LDA on the data 
4. predict topics inside the document.


    The latent Dirichlet allocation (LDA) is a common
    method for topic modelling, based
    on the assumption that each document in a corpus is
    composed by one or more
    hidden topics, and each topic is supported by a number of
    words. The process is to
    find these hidden topics and their supporting words by 
    maximising the posterior
    probability of the whole corpus given the topics and
    words.



In [ ]:
pip install pyLDAvis

In [2]:
import nltk
from nltk.corpus import stopwords
from nltk.stem.snowball import SnowballStemmer
from gensim import models, corpora


In [3]:
documents = [
  """
  Artificial intelligence (AI), sometimes called machine
  intelligence, is intelligence demonstrated by machines, unlike
  the natural intelligence displayed by humans and animals. Leading
  AI textbooks define the field as the study of "intelligent
  agents": any device that perceives its environment and takes
  actions that maximize its chance of successfully achieving its
  goals. Colloquially, the term "artificial intelligence" is often
  used to describe machines (or computers) that mimic "cognitive"
  functions that humans associate with the human mind, such
  as "learning" and "problem solving".
  """,
  """
  Association football, more commonly known as football or
  soccer, is a team sport played with a spherical ball between
  two teams of 11 players. It is played by approximately 250
  million players in over 200 countries and dependencies, making it
  the world's most popular sport. The game is played on a
  rectangular field called a pitch with a goal at each end. The
  object of the game is to outscore the opposition by moving the
  ball beyond the goal line into the opposing goal. The team with
  the higher number of goals wins the game.  
  """
]

In [4]:
# Clean the data by using stemming and stopwords removal
nltk.download('stopwords')
stemmer = SnowballStemmer('english')
stop_words = stopwords.words('english')
texts = [
  [stemmer.stem(word) for word in document.lower().split() if word not in stop_words]
  for document in documents
  ]

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


In [5]:
# Create a dictionary from the words
dictionary = corpora.Dictionary(texts)

# Create a document-term matrix
doc_term_mat = [dictionary.doc2bow(text) for text in texts]

# Generate the LDA model 
num_topics = 2
ldamodel = models.ldamodel.LdaModel(doc_term_mat, 
        num_topics=num_topics, id2word=dictionary, passes=25)


In [6]:
num_words = 5
for i in range(num_topics):
  print(ldamodel.print_topic(i, topn=num_words))

print('\nTop ' + str(num_words) + ' contributing words to each topic:')
for item in ldamodel.print_topics(num_topics=num_topics, num_words=num_words):
    print('\nTopic', item[0])
    list_of_strings = item[1].split(' + ')
    for text in list_of_strings:
        details = text.split('*')
        print("%-12s:%0.2f%%" %(details[1], 100*float(details[0])))


0.035*"human" + 0.035*"intellig" + 0.025*"machin" + 0.015*"call" + 0.015*"associ"
0.036*"goal" + 0.036*"play" + 0.036*"team" + 0.026*"game" + 0.026*"ball"

Top 5 contributing words to each topic:

Topic 0
"human"     :3.50%
"intellig"  :3.50%
"machin"    :2.50%
"call"      :1.50%
"associ"    :1.50%

Topic 1
"goal"      :3.60%
"play"      :3.60%
"team"      :3.60%
"game"      :2.60%
"ball"      :2.60%


In [12]:
new_docs = [
  """
  footabll, goal is good. i love football.yesterday's game was good.
  AI is is eqally intresting as football. we can try to predict game output
  using AI algorithms.
  """
]

new_texts = [
  [stemmer.stem(word) for word in document.lower().split() if word not in stop_words]
  for document in new_docs
  ]
new_doc_term_mat = [dictionary.doc2bow(text) for text in new_texts]

vector = ldamodel[new_doc_term_mat]
print(vector[0])


[(0, 0.45892093), (1, 0.5410791)]


In [13]:
import pyLDAvis.gensim_models
pyLDAvis.enable_notebook()
vis = pyLDAvis.gensim_models.prepare(ldamodel,doc_term_mat,dictionary,mds="mmds",R=5)
vis


/usr/local/lib/python3.8/dist-packages/pyLDAvis/_prepare.py:246: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  default_term_info = default_term_info.sort_values(


PreparedData(topic_coordinates=              x         y  topics  cluster       Freq
topic                                                
0     -0.022433  0.070815       1        1  50.940799
1      0.022433 -0.070815       2        1  49.059201, topic_info=        Term      Freq     Total Category  logprob  loglift
63      goal  2.000000  2.000000  Default   5.0000   5.0000
77      play  2.000000  2.000000  Default   4.0000   4.0000
86      team  2.000000  2.000000  Default   3.0000   3.0000
27     human  2.000000  2.000000  Default   2.0000   2.0000
28  intellig  2.000000  2.000000  Default   1.0000   1.0000
27     human  1.908757  2.177132   Topic1  -3.3425   0.5429
28  intellig  1.908741  2.177132   Topic1  -3.3425   0.5429
32    machin  1.363334  1.631676   Topic1  -3.6790   0.4948
21     devic  0.817927  1.086221   Topic1  -4.1899   0.3908
36     mind,  0.817924  1.086221   Topic1  -4.1899   0.3908
14      call  0.818121  1.622305   Topic1  -4.1897  -0.0101
13    associ  0.818119  1.622305   Topic1  -4.1897  -0.0101
63      goal  1.876024  2.149017   Topic2  -3.3221   0.5763
77      play  1.876023  2.149017   Topic2  -3.3221   0.5763
86      team  1.876022  2.149017   Topic2  -3.3221   0.5763
61      game  1.339955  1.612935   Topic2  -3.6587   0.5267
53      ball  1.339952  1.612935   Topic2  -3.6587   0.5267, token_table=      Topic      Freq      Term
term                           
13        1  0.616407    associ
13        2  0.616407    associ
53        2  0.619988      ball
14        1  0.616407      call
14        2  0.616407      call
21        1  0.920623     devic
61        2  0.619988      game
63        2  0.930658      goal
27        1  0.918640     human
28        1  0.918640  intellig
32        1  0.612867    machin
36        1  0.920623     mind,
77        2  0.930658      play
86        2  0.930658      team, R=5, lambda_step=0.01, plot_opts={'xlab': 'PC1', 'ylab': 'PC2'}, topic_order=[1, 2])

##**Discussion**

 * LDA is a really effective way to to get number of topic inside a big document.

* It is also important to remove the stop words from the documents.
Depending on the language we use stop words package to remove it. Here we have used SnowballStemmer for the stop words.
Also, reducing words to its base form, gives better output.


    eg. reduce, reduces , reducing, reduced --> reduce
    here all form are reduced to a simpler form "reduce".


 * also once a LDA model is ready, we can pass new document 
 to it and can get the probability based result, based on the topic.


    eg: we can get with what probability,new document 
    belongs to already defined category.

    * sample result of new doc being passed:--
              [(0, 0.45892093), (1, 0.5410791)]
                * for topic o probability is 45%
                * and for topic 1 probaility is 54%

